iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java

```diff
@@ -0,0 +1,209 @@
+    }
+
+    if (arguments[0].getCategory() != ObjectInspector.Category.PRIMITIVE) {
+      throw new UDFArgumentException(
+          "ICEBERG_BUCKET first argument takes primitive types, got " + argumentOI.getTypeName());
+    }
+    argumentOI = (PrimitiveObjectInspector) arguments[0];
+
+    PrimitiveObjectInspector.PrimitiveCategory inputType = argumentOI.getPrimitiveCategory();
+    ObjectInspector outputOI = null;
+    switch (inputType) {
+      case CHAR:
+      case VARCHAR:
+      case STRING:
+        icebergType = Types.StringType.get();
+        stringConverter = new PrimitiveObjectInspectorConverter.StringConverter(argumentOI);
+        break;
+
+      case BINARY:
+        icebergType = Types.BinaryType.get();
+        binaryConverter = new PrimitiveObjectInspectorConverter.BinaryConverter(argumentOI,
+            PrimitiveObjectInspectorFactory.writableBinaryObjectInspector);
+        break;
+
+      case INT:
+        icebergType = Types.IntegerType.get();
+        intConverter = new PrimitiveObjectInspectorConverter.IntConverter(argumentOI,
+            PrimitiveObjectInspectorFactory.writableIntObjectInspector);
+        break;
+
+      case LONG:
+        icebergType = Types.LongType.get();
+        longConverter = new PrimitiveObjectInspectorConverter.LongConverter(argumentOI,
+            PrimitiveObjectInspectorFactory.writableLongObjectInspector);
+        break;
+
+      case DECIMAL:
+        DecimalTypeInfo decimalTypeInfo = (DecimalTypeInfo) TypeInfoUtils.getTypeInfoFromObjectInspector(argumentOI);
+        icebergType = Types.DecimalType.of(decimalTypeInfo.getPrecision(), decimalTypeInfo.getScale());
+
+        decimalConverter = new PrimitiveObjectInspectorConverter.HiveDecimalConverter(argumentOI,
+            PrimitiveObjectInspectorFactory.writableHiveDecimalObjectInspector);
+        break;
+
+      case FLOAT:
+        icebergType = Types.FloatType.get();
+        floatConverter = new PrimitiveObjectInspectorConverter.FloatConverter(argumentOI,
+            PrimitiveObjectInspectorFactory.writableFloatObjectInspector);
+        break;
+
+      case DOUBLE:
+        icebergType = Types.DoubleType.get();
+        doubleConverter = new PrimitiveObjectInspectorConverter.DoubleConverter(argumentOI,
+            PrimitiveObjectInspectorFactory.writableDoubleObjectInspector);
+        break;
+
+
+      default:
+        throw new UDFArgumentException(
+            " ICEBERG_BUCKET() only takes STRING/CHAR/VARCHAR/BINARY/INT/LONG/DECIMAL/FLOAT/DOUBLE" +
+                " types as first argument, got " + inputType);
+    }
+
+    argumentOI = (PrimitiveObjectInspector) arguments[1];
+    if (!isValid2ndArgumentType(arguments[1])) {
+      throw new UDFArgumentTypeException(1,
+          "ICEBERG_BUCKET() second argument only takes an int types, got " + arguments[1].getTypeName());
+    }
+    numBuckets = ((WritableConstantIntObjectInspector) argumentOI).getWritableConstantValue().get();
```

Comment:
Question: It might be just me, but instead of `isValid2ndArgumentType` I would create a `private int getNumBuckets(ObjectInspector) throws UDFArgumentException` method

Reply:
makes sense